### ISSUE TITLE
Possible errors in gt_data.csv

### ISSUE THREAD

**Issue by stuarteberg** [funkelab/drosophila_neurotransmitters#6]
While investigating possible issues with the MaleCNS NT groundtruth, I noticed that there have been many changes to the `gt_data.csv` file in this repo.  I happened to compare the version from 2024-12-05 with the latest version.

Aside from new sources of evidence, it seems that many of the existing entries have changed in the NT expression columns.

Here's a comparison of the old (2024-12) and new (2025-06) versions.  This only includes rows in which the source and cell type match, but the neurotransmitter expression seems to differ in at least one column.  (For now, I'm ignoring cases where the cell types were renamed or a reference was added or removed.)

https://docs.google.com/spreadsheets/d/1lY7sjBfykWz8seVQEnJxl92TuyKyCAL-0lvkfn8P6PA/edit?gid=1528009826#gid=1528009826

Based on some spot-checks performed by Aljoscha Nern, we think many of the changes are likely mistaken (e.g. neurotransmitter expression flipping from negative to positive.)

Would you mind having a look and telling us what you think, @alexanderbates?

If it would help, I can re-run my comparison using different snapshots of `gt_data.csv` than the ones mentioned above.

Thanks!

**Comment by aljoschanern** [comment]
To add to the comment above: The current gt version (2025-06) also contains several new entries which I think are not supported by the cited references.  For example, DmDRA2 and Dm16 data attributed to Davis et al 2020. These cell types were not part of the Davis et al study.  There several other examples like this.

**Comment by alexanderbates** [comment]
Thank you both, sorry for the late response. Thank you for documenting the issue so well.

I think what happened here is that we changed visual types from the Jefferis lab curation (https://github.com/flyconnectome/ol_annotations) to the Princeton curation (i.e. https://www.nature.com/articles/s41586-024-07981-1, what is on FlyWire Codex) for FAFB. I think I might go back to the Jefferis lab curation.

We did this in the underlying database in which I record NT information by neuron ID across connectomes (there's arguments for and against having a database by cell type for this, and for better or worse I chose this path early on).  In the switch, the same IDs changed cell type identities, but erroneously kept their old NT labels, which shuld have been wiped/ported with the change in cell type identity.

In these cases (mis-match between Jefferis lab and Princeton annotation) I should remove the per-ID NT label, and then re-calculate this per-cell type file to be accurate. 

I will work on that today/tomorrow, I apologise for any inconvenience this caused. We used the prior version for the recent BANC predictions, I think (though because we are missing much of the optic lobes at the moment, this error may not effect us much).

**Comment by stuarteberg** [comment]
Thank you, @alexanderbates for looking into this!

>I removed some of the negative results

Ah.  Okay, in case it's useful, I've tweaked the colors in the comparison spreadsheet to differentiate between cases where a negative was removed vs. cases where a 0 changed to something else.

>Do you use the negatives?

No, except that I do discard any types which seem to have *conflicting* evidence (one study +1, another -1) from our training data.  But there aren't many of those, and I have no strong opinion on whether the negatives should be included in the table or not.

>For optic lobe types:

Okay, I see that most of the transitions from 0->1 or -1->1 are in the optic lobe types.  There also seem to be a few in the larval cell types (which I don't use).  But there is at least one example in the adult midbrain: `CB.FB6E2`.

>I apologise for any inconvenience this caused

Not to worry, our NT predictor was trained on the old data.  However, we're looking into how we might improve it, so I was looking at what could be gained by updating our ground truth, which is when I noticed these oddities.

If you're able to sort things out, we will update our ground truth using your updated table.  An additional benefit will be that the names in our database may be more consistent with the newer types than it was when we first trained, at least in some cases (e.g. `Dm3q`).  If I switch to the newer table, we might avoid missing some types in our ground truth (if we retrain our model).

**Comment by alexanderbates** [comment]
Great!

See most recent push.

I spent the day fixing up these issues, and spotting more, mainly from the optic lobe type confusion. This accounts for the mis-attributions that Aljoscha noted as well. In small number of cases, a wrong label was applied. I have also reviewed the GT more broadly, and added and dropped some labels. 

(`CB.FB6E2` (also called FB6H) had legitimate changes for example, it is a super weird cell (one of the 'dFB' cell types noted in the literature) that recent work has assigned to have glutamate, dopamine and - possibly - serotonin as well. It may also form unusual gap junctions. I would not include in GT)

In addition, Kevin Delgado and I found that for BANC, omitting certain cell types with legitimate ground truth wetlab data helps. We don't always know why - the wetlab study could be wrong, the matching to EM could be wrong or there could be confounding transmission (I think the latter is most likely). Another good idea is to remove all motor neurons. They are known to be glutamatergic but they do not have legitimate outputs in the main CNS (usually) and so we were picking up junk presynapses for 'glutamate' there. Kenyon cells may or may not be weird for you (I am interested to know what happens for you, I am sure Diane is as well!) - and could be removed or treated as their own special class of synapse (?). 

I have added a file that gives you our exclusions, find it here: `gt_sources/banc/banc_cell_types_removed_for_nt.csv`. I recommend using! But our cell type names may not all map onto yours (I cannot see maleCNS data so I don't know).

These

**Comment by stuarteberg** [comment]
Thank you so much for working on this and for the helpful explanations, @alexanderbates!  We're lucky to have you as a contributor.

I've compiled [a new comparison of the old (2024-09) vs. new (2025-08) `gt_data.csv` contents](https://docs.google.com/spreadsheets/d/1lY7sjBfykWz8seVQEnJxl92TuyKyCAL-0lvkfn8P6PA/edit?gid=1922483604#gid=1922483604).  This time I just erased all negative entries before the comparison.  (This view isn't perfect since in some cases I may not have matched the reference names perfectly. You may see some pairs of removed/added rows instead of changed/changed rows, or vice-versa.)

Thank you for sharing your advice on which cells to exclude for training.  I clearly didn't put enough thought into that before translating your original table for Diane to use on our data.  I'm now in the process of compiling a comparison of the cells we used previously vs. the ones would use now if we were to re-train the model with (1) our latest cell type annotations and (2) your refined ground truth table.  The near-perfect accuracy you report for BANC gives me hope that we have room for improvement.  BTW, Congratulations on posting the BANC preprint!

**Comment by aljoschanern** [comment]
@alexanderbates Thanks the explanations and updated table !  I may have a few more comments on minor issue I noticed, I'll provide these later.   One request: Would it be possible to add a column with the cell type names used in the cited source studies directly to the gt table ?  I am assuming this information already exists somewhere.  Since many of the issues discussed above are related to cell type (name) matching, I think this would greatly improve the reliability of the gt table as a resource.  (Obviously, there are cases  for which there is no original name e.g. for lineage-based assignments.)

**Comment by alexanderbates** [comment]
Thank you both, and also for your kind words there Stuart. The google sheet is a very good way of seeing the changes - skimming through, these changes should all be intentional and if you would like an explanation for any I am happy to do so. Some GT has been changing in dialogue with Kevin, as early BANC results prompted me to rethink some of my calls. In addition the Lacin GT is based on hemilineage, and then selecting neurons types from MANC with good predictions for that hemilineage's wet-lab assigned GT, dropping those types that lack them (in case there is hidden within-lineage diversity), hence why many of those look to have changed.

The BANC results look 'near perfect' as you say, but they did not look good for a long time. With Kevin, we iterated a lot and dropped neurons from the train set that kept being assigned the wrong transmitter if put in the test set, and for which if we thought about it the wetlab NT assignment or neuron matching might have been shaky. We also had some neurons that should diversity in the predicted synapse across their arbours (we think for image artefact reasons or merge errors) and we also dropped those from GT. That's what built our exclusion list (btw, I just updated it so please pull again!). Kevin made other changes on how he sampled the GT, and you may wish to reach out to him for details.   

You'll note that the preprint barely uses them, and this is partly because we got them at the eleventh hour, and partly because I have not yet had the time to deeply assess them. Some issues that also exist in FAFB, persist (maybe slightly lesser), e.g. dopamine and serotonin predictions are about 10x what we expect, and this is likely due to peptide transmitters in some neurons confusing the network. It just so happens that all the serotonin GT are for neurons that also co-release neuropeptides, we think. So outside of the GT used (which isn't all of it because BANC probably still has fewer neurons matched than maleCNS), the BANC predictions perform as well as FAFB (which is good!) but not as good as our supplement makes it look I think. I will try to prepare another supplement for the NT prediction specifically that can show some caveats.

Aljoscha that's a great idea. Actually the intention was just to list other names as separate rows (this already happens a little bit) but maybe an explicit name mapping would be better. This information is a little clumsily curated, it exists in my database as a free form 'notes' column so some regex is needed to get relevant other names. If you have any corrections or name mappings you care most about you wish to send, I will start by integrating those! Some mappings already exist here: `extdata/cell_type_cross_matching.csv`

I cannot see maleCNS data and so there's no attempt to actually integrate with those names. I assume they inherit from Hemibrain/FAFB a lot.

**Comment by alexanderbates** [comment]
Thank you Aljoscha! This looks very useful, I aim to integrate it on the weekend. Responding quickly, I had avoided giving 'cell_type_[dataset]' and wanted the cell type column just to repeat for different terms, with the same data in the other rows where possible. The logic being that cell type names in connectome change over time, but rarely is the same name used for a different cell type (at least intentionally) so the user wants to search against the whole list. However, what you suggest may well be simpler, it would just take me a bit of time to execute. I'll check in on the references you mention, but all your actions sound sensible to me.

Of interest to @stuarteberg as well: we got BANC NT predictions late in the day, but added them to the preprint (and not the analysis for time reasons) so that the community can start to use them. One fairly large 'bug' I just noticed is that we do not have any dopamine predictions in the VNC. We know there is a bunch of dopamine in the VNC from wetlab work, so this is a surprise and might suggest we have overlearned or the solely brain-base dopamine GT is an issue. 

Do you have wetlab assignments for any VNC dopamine neurons? Or serotonin for that matter?

**Comment by alexanderbates** [comment]
Hmm I think I found some [here](https://physoc.onlinelibrary.wiley.com/doi/full/10.14814/phy2.12322), seemingly DNge047. This may be a very useful piece of GT, if you are having the same issue.

**Comment by aljoschanern** [comment]
Thanks, Alex ! You are right that the same name is rarely used for different cell types, but there are exceptions, including for some cell types in gt_data (the examples I am thinking of are OL cell type for which the issue should be addressed in the modified version I shared).

I found some other potential errors:  I looked at some of the Wolff et al 2024 entries in gt_data and noticed  cases for which I can't find any entries in the paper. A few examples:  FB5X, FB1J, FB8G (there are more). Maybe the same (or a similar issue) as mentioned above (and then resolved) for some optic lobe cell types ?
